src/runtime.js

```diff
@@ -195,7 +196,10 @@ function adoptShared(global, opts, callback) {
   /** @const {!./service/extensions-impl.Extensions} */
   const extensions = extensionsFor(global);
   installRuntimeServices(global);
-  stubLegacyElements(global);
+
+  if (!opts.isCriticalPath) {
+    stubLegacyElements(global);
```

Comment:
This is related to the changes in `amp-shadow.js`

Our problem was that currently `amp-shadow.js` assumes that there won't be any AMP component in the DOM while intializing the runtime in shadow mode. That's why is safe to call `installBuiltins` and `stubElements` although at that point there's still any Resources Service available.

Our solution was to switch execution flow: call `adoptShadowMode` (where an `AmpDoc` and its services are installed for document ), and then initialize builtin AMP components

However, with your concept of `AmpDocShell` perhaps there's a better solution for this. We can remove the call `manager.attachShadowDoc` inside the method `adoptShadowMode` in `runtime.js`, and in amp-shadow.js we can initialize the `AmpDocShell` and install it's services, and follow the current execution flow. This mean we could also remove this change